File: examples/rust/basic_types.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright © 2025 Au-Zone Technologies. All Rights Reserved.

//! Basic Types Example
//!
//! Demonstrates EdgeFirst Schemas CDR serialization patterns:
//! - CdrFixed types: compile-time sized, value semantics (Time, Vector3, Pose, ColorRGBA)
//! - Buffer-backed types: zero-copy views over CDR byte buffers (Header, Image)
//!
//! The library uses CDR1 Little-Endian wire format, compatible with ROS 2 DDS.

use edgefirst_schemas::builtin_interfaces::Time;
use edgefirst_schemas::cdr::{decode_fixed, encode_fixed};
use edgefirst_schemas::geometry_msgs::{Point, Pose, Quaternion, Transform, Vector3};
use edgefirst_schemas::sensor_msgs::Image;
use edgefirst_schemas::std_msgs::{ColorRGBA, Header};

/// CdrFixed types: fixed-size structs serialized with encode_fixed/decode_fixed.
fn example_fixed_types() {
    println!("=== CdrFixed Types ===\n");

    // Time — 8 bytes on the wire (i32 sec + u32 nanosec)
    let time = Time::new(1234567890, 123456789);
    let bytes = encode_fixed(&time).unwrap();
    let decoded: Time = decode_fixed(&bytes).unwrap();
    assert_eq!(time, decoded);
    println!("Time:       {}.{:09}s  ({} CDR bytes)", time.sec, time.nanosec, bytes.len());

    // Vector3 — 24 bytes (3 × f64)
    let vel = Vector3 { x: 1.5, y: 2.0, z: 0.5 };
    let bytes = encode_fixed(&vel).unwrap();
    let decoded: Vector3 = decode_fixed(&bytes).unwrap();
    assert_eq!(vel, decoded);
    let mag = (vel.x * vel.x + vel.y * vel.y + vel.z * vel.z).sqrt();
    println!("Vector3:    ({}, {}, {})  mag={:.3}  ({} bytes)", vel.x, vel.y, vel.z, mag, bytes.len());

    // Pose — 56 bytes (Point 24 + Quaternion 32)
    let pose = Pose {
        position: Point { x: 1.0, y: 2.0, z: 0.5 },
        orientation: Quaternion { x: 0.0, y: 0.0, z: 0.707, w: 0.707 },
    };
    let bytes = encode_fixed(&pose).unwrap();
    let decoded: Pose = decode_fixed(&bytes).unwrap();
    assert_eq!(pose, decoded);
    println!("Pose:       pos=({},{},{}) quat=({},{},{},{})  ({} bytes)",
        pose.position.x, pose.position.y, pose.position.z,
        pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w,
        bytes.len());

    // Transform — 56 bytes
    let tf = Transform {
        translation: Vector3 { x: 1.0, y: 2.0, z: 0.5 },
        rotation: Quaternion { x: 0.0, y: 0.0, z: 0.0, w: 1.0 },
    };
    let bytes = encode_fixed(&tf).unwrap();
    let decoded: Transform = decode_fixed(&bytes).unwrap();
    assert_eq!(tf, decoded);
    println!("Transform:  ({} bytes)", bytes.len());

    // ColorRGBA — 16 bytes (4 × f32)
    let color = ColorRGBA { r: 1.0, g: 0.0, b: 0.0, a: 1.0 };
    let bytes = encode_fixed(&color).unwrap();
    let decoded: ColorRGBA = decode_fixed(&bytes).unwrap();
    assert_eq!(color, decoded);
    println!("ColorRGBA:  r={} g={} b={} a={}  ({} bytes)\n", color.r, color.g, color.b, color.a, bytes.len());
}

/// Buffer-backed types: generic over B: AsRef<[u8]>, wrapping a CDR byte buffer.
/// Construction scans the buffer once to build a small offset table for O(1) field access.
fn example_buffer_backed_types() {
    println!("=== Buffer-Backed Types ===\n");

    // Header<Vec<u8>> — owned buffer, constructed with new()
    let stamp = Time::new(1234567890, 123456789);
    let header = Header::new(stamp, "camera_optical_frame").unwrap();
    println!("Header:     stamp={}.{:09}  frame_id=\"{}\"  ({} CDR bytes)",
        header.stamp().sec, header.stamp().nanosec, header.frame_id(), header.cdr_size());

    // Round-trip: to_cdr() serializes, from_cdr() deserializes
    let cdr_bytes = header.to_cdr();
    let decoded = Header::from_cdr(cdr_bytes).unwrap();
    assert_eq!(decoded.stamp(), stamp);
    assert_eq!(decoded.frame_id(), "camera_optical_frame");
    println!("            round-trip OK");

    // Zero-copy borrow: from_cdr(&[u8]) borrows the buffer without copying
    let owned_bytes = header.to_cdr();
    let borrowed = Header::from_cdr(owned_bytes.as_slice()).unwrap();
    assert_eq!(borrowed.frame_id(), "camera_optical_frame");
    println!("            zero-copy borrow OK");

    // Image<Vec<u8>> — larger buffer-backed type
    let pixel_data = vec![128u8; 640 * 480 * 3]; // VGA RGB8
    let img = Image::new(stamp, "camera", 480, 640, "rgb8", 0, 640 * 3, &pixel_data).unwrap();
    println!("\nImage:      {}x{} encoding=\"{}\"  data={} bytes  CDR={} bytes",
        img.width(), img.height(), img.encoding(), img.data().len(), img.cdr_size());

    // Round-trip
    let cdr_bytes = img.to_cdr();
    let decoded = Image::from_cdr(cdr_bytes).unwrap();
    assert_eq!(decoded.width(), 640);
    assert_eq!(decoded.height(), 480);
    assert_eq!(decoded.encoding(), "rgb8");
    assert_eq!(decoded.data().len(), 640 * 480 * 3);
    println!("            round-trip OK\n");
}

/// Mutation: buffer-backed types with mutable/owned buffers support in-place field updates.
fn example_mutation() {
    println!("=== Mutation ===\n");

    let mut header = Header::new(Time::new(0, 0), "test").unwrap();
    println!("Before:     stamp={}.{:09}", header.stamp().sec, header.stamp().nanosec);

    header.set_stamp(Time::new(42, 123)).unwrap();
    println!("After:      stamp={}.{:09}\n", header.stamp().sec, header.stamp().nanosec);
    assert_eq!(header.stamp(), Time::new(42, 123));
}

fn main() {
    println!("EdgeFirst Schemas - CDR Serialization Examples");
    println!("===============================================\n");

    example_fixed_types();
    example_buffer_backed_types();
    example_mutation();

    println!("===============================================");
    println!("All examples completed successfully!");
}

File: examples/python/basic_types.py
#!/usr/bin/env python3
"""
Basic Types Example

SPDX-License-Identifier: Apache-2.0
Copyright © 2025 Au-Zone Technologies. All Rights Reserved.

Demonstrates fundamental ROS2 message types including:
- builtin_interfaces (Time, Duration)
- std_msgs (Header, ColorRGBA)
- geometry_msgs (Vector3, Point, Quaternion, Pose)
"""

from edgefirst.schemas.builtin_interfaces import Duration, Time
from edgefirst.schemas.geometry_msgs import Point, Pose, Quaternion, Vector3
from edgefirst.schemas.std_msgs import ColorRGBA, Header


def example_time() -> None:
    """Demonstrate Time message."""
    print("=== Example: Time ===")

    # Create a timestamp
    time = Time(sec=1234567890, nanosec=123456789)

    print(f"Time: {time.sec}.{time.nanosec:09d} seconds")
    print(f"Time (repr): {time!r}\n")


def example_duration() -> None:
    """Demonstrate Duration message."""
    print("=== Example: Duration ===")

    # Create a duration (5.5 seconds)
    duration = Duration(sec=5, nanosec=500_000_000)

    total_ns = duration.sec * 1_000_000_000 + duration.nanosec
    print(f"Duration: {total_ns / 1e9} seconds ({total_ns} nanoseconds)\n")


def example_header() -> None:
    """Demonstrate Header message."""
    print("=== Example: Header ===")

    # Create a header with timestamp and frame ID
    header = Header(
        stamp=Time(sec=1234567890, nanosec=123456789),
        frame_id="camera_optical_frame"
    )

    print("Header:")
    print(f"  timestamp: {header.stamp.sec}.{header.stamp.nanosec:09d}")
    print(f"  frame_id: {header.frame_id}\n")

    # Default initialization
    default_header = Header()
    print(f"Default header: {default_header!r}\n")


def example_color() -> None:
    """Demonstrate ColorRGBA message."""
    print("=== Example: ColorRGBA ===")

    # Create colors
    red = ColorRGBA(r=1.0, g=0.0, b=0.0, a=1.0)
    transparent_blue = ColorRGBA(r=0.0, g=0.0, b=1.0, a=0.5)

    print(f"Red: {red!r}")
    print(f"Transparent blue: {transparent_blue!r}\n")


def example_vector3() -> None:
    """Demonstrate Vector3 message."""
    print("=== Example: Vector3 ===")

    velocity = Vector3(x=1.5, y=2.0, z=0.5)

    # Calculate magnitude
    magnitude = (velocity.x**2 + velocity.y**2 + velocity.z**2) ** 0.5

    print(f"Velocity vector: ({velocity.x}, {velocity.y}, {velocity.z})")
    print(f"Magnitude: {magnitude:.3f}\n")


def example_point() -> None:
    """Demonstrate Point message."""
    print("=== Example: Point ===")

    origin = Point(x=0.0, y=0.0, z=0.0)
    target = Point(x=10.0, y=5.0, z=2.0)

    # Calculate distance
    dx = target.x - origin.x
    dy = target.y - origin.y
    dz = target.z - origin.z
    distance = (dx * dx + dy * dy + dz * dz) ** 0.5

    print(f"Origin: {origin!r}")
    print(f"Target: {target!r}")
    print(f"Distance: {distance:.3f}\n")


def example_quaternion() -> None:
    """Demonstrate Quaternion message."""
    print("=== Example: Quaternion ===")

    # Identity quaternion (no rotation)
    identity = Quaternion(x=0.0, y=0.0, z=0.0, w=1.0)

    # 90 degree rotation around Z axis
    rotation_z_90 = Quaternion(x=0.0, y=0.0, z=0.707, w=0.707)

    print(f"Identity rotation: {identity!r}")
    print(f"90° Z rotation: {rotation_z_90!r}\n")


def example_pose() -> None:
    """Demonstrate Pose message."""
    print("=== Example: Pose ===")

    pose = Pose(
        position=Point(x=1.0, y=2.0, z=0.5),
        orientation=Quaternion(x=0.0, y=0.0, z=0.0, w=1.0)
    )

    print("Pose:")
    print(f"  position: ({pose.position.x}, {pose.position.y}, {pose.position.z})")
    print(f"  orientation: ({pose.orientation.x}, {pose.orientation.y}, "
          f"{pose.orientation.z}, {pose.orientation.w})\n")


def example_serialization() -> None:
    """Demonstrate CDR serialization."""
    print("=== Example: Serialization ===")

    header = Header(
        stamp=Time(sec=1234567890, nanosec=123456789),
        frame_id="test_frame"
    )

    # Serialize to CDR
    bytes_data = header.serialize()
    print(f"Serialized header to {len(bytes_data)} bytes")

    # Deserialize from CDR
    decoded = Header.deserialize(bytes_data)

    # Verify round-trip
    assert decoded.stamp.sec == header.stamp.sec
    assert decoded.stamp.nanosec == header.stamp.nanosec
    assert decoded.frame_id == header.frame_id

    print("Round-trip successful!")
    print(f"  Original:     {header!r}")
    print(f"  Deserialized: {decoded!r}\n")


def main() -> None:
    """Run all examples."""
    print("EdgeFirst Schemas - Basic Types Examples")
    print("=" * 50 + "\n")

    example_time()
    example_duration()
    example_header()
    example_color()
    example_vector3()
    example_point()
    example_quaternion()
    example_pose()
    example_serialization()

    print("=" * 50)
    print("All examples completed successfully!")


if __name__ == "__main__":
    main()
